hs_labels/models.py

```diff
@@ -0,0 +1,876 @@
+
+    *This exception is a generic base class and is never directly raised.*
+    See subclass HSLUsageException for details.
+    """
+    pass
+
+class HSLUsageException(HSLException):
+    """
+    Exception class for parameter usage problems.
+
+    This exception is raised when a routine is called with invalid parameters.
+    This includes references to non-existent resources, groups, and users.
+
+    This is typically a programming error and should be entered into
+    issue tracker and resolved.
+
+    *Catching this exception is not recommended.*
+    """
+    def __str__(self):
+        return repr("HS Usage Exception: " + self.message)
+
+    pass
+
+
+####################################
+####################################
+# The labeling system allows
+# a) labeling of resources.
+# b) assigning folders to resources.
+# c) marking public resources to be "of interest"
+####################################
+####################################
+
+
+####################################
+# privileges are numeric codes 1-4
+####################################
+class LabelCodes(object):
+    """
+    Label codes describe the kinds of labels that a user can create for a thing:
+
+        * 1 or LabelCodes.LABEL:
+            a regular label for the object.
+
+        * 2 or LabelCodes.FOLDER:
+            a folder for the object.
+
+        * 3 or LabelCodes.FAVORITE:
+            marks the object as a favorite.
+
+        * 4 or LabelCodes.MINE:
+            marks the object to be included in my resources
+
+        * 5 or LabelCodes.SAVEDLABEL:
+            creates a label that can be used later for labeling a resource
+
+    """
+    LABEL = 1
+    FOLDER = 2
+    FAVORITE = 3
+    MINE = 4
+    SAVEDLABEL = 5
+    LABEL_CHOICES = (
+        (LABEL, 'Label'),
+        (FOLDER, 'Folder'),
+        (FAVORITE, 'Favorite'),
+        (MINE, 'Mine'),
+        (SAVEDLABEL, 'Saved Label')
+    )
+
+####################################
+# user access to resources
+####################################
+
+
+class UserResourceLabels(models.Model):
+    """ Labels of a user for a resource
+
+    This model stores labels of an individual user, like an access
+    control list; There are several kinds of labels documented in LabelCodes.
+    These are similar in implementation but differ in semantics.
+    """
+
+    kind = models.IntegerField(choices=LabelCodes.LABEL_CHOICES,
+                               editable=False,
+                               default=LabelCodes.LABEL)
+    # provenance of label start
+    start = models.DateTimeField(editable=False, auto_now=True)
+
+    # if end is non-null, then the label is no longer valid. Not absolutely sure we need this.
+    # end  = models.DateTimeField(editable=False, null=True, auto_now=False)
+
+    # Utilize an underlying class of User to hold labels for resources.
+    # This creates an implicit query User.ulabels that represents all labels
+    # for a user.
```

Comment:
Again, I must protest omitting comments that I need to remember how my class works!